### ISSUE TITLE
ENH: Speedup to overlay/identity

### ISSUE THREAD

**Issue by theroggy** [geopandas/geopandas#3295]
While reading the code of `overlay`, I noticed that for "identity", "union" is calculated (= "intersection" + "symmetric_difference") and then part of the result is removed. This could be replaced by "intersection" + "difference", which should be faster?

**Comment by BMarques** [comment]
Do you mean that doing something like this 

# Perform the intersection to get overlapping features
intersections = gpd.overlay(input_gdf, identity_gdf, how='intersection')

# Perform the difference to get non-overlapping features from input_gdf
difference = gpd.overlay(input_gdf, identity_gdf, how='difference')

# Combine the results
result_gdf = intersections.append(difference, ignore_index=True)

will yield the same results, but quicker?

**Comment by martinfleis** [comment]
I think that it was directed towards internal code that lives in https://github.com/geopandas/geopandas/blob/e82949886a9937b39d1d5dce087d1ad5ccaf01d4/geopandas/tools/overlay.py#L326-L328

where `_overlay_union` consists of

https://github.com/geopandas/geopandas/blob/e82949886a9937b39d1d5dce087d1ad5ccaf01d4/geopandas/tools/overlay.py#L140-L141

@theroggy I suppose it is worth trying.

**Comment by BMarques** [comment]
Sorry, I was aware of what he was talking about, I kinda super imposed because I'm stuck currently with the slow performance of overlay identity (takes more than 3 hours in my use case), and I was wondering if doing what I said manually also did the same thing, based on his comment.

**Comment by theroggy** [comment]
> Sorry, I was aware of what he was talking about, I kinda super imposed because I'm stuck currently with the slow performance of overlay identity (takes more than 3 hours in my use case), and I was wondering if doing what I said manually also did the same thing, based on his comment. Will stop "hijacking" the issue now.

Yes, doing this "manually" should give the same performance improvement... if it does improve performance. So, if you would try it on your use case... please report back here because if it is faster it would be a confirmation that it is a good idea to change the internal implementation. If it would not be faster, it is only a theoretical improvement and there is no use in changing it.

**Comment by Bimbam360** [comment]
Chiming in as @BMarques never replied.
In my example I've just replaced:

```
overlay_start = time.perf_counter()
gdf_layers['union_cadastres'] = gpd.overlay(
	gdf_layers['cadastre'],
	gdf_layers['union'],
	how="identity",
	keep_geom_type=False
).explode(index_parts=False)
gdf_layers['union_cadastres'] = gdf_layers['union_cadastres'][gdf_layers['union_cadastres'].geometry.geom_type == "Polygon"]
print(f"Overlay Identity took {time.perf_counter() - overlay_start:.2f} seconds")
```

with
```
overlay_start = time.perf_counter()
intersections = gpd.overlay(gdf_layers['cadastre'], gdf_layers['union'], how='intersection', keep_geom_type=False)
difference = gpd.overlay(gdf_layers['cadastre'], gdf_layers['union'], how='difference', keep_geom_type=False)
gdf_layers['union_cadastres_test'] = pd.concat([intersections, difference], ignore_index=True).explode(index_parts=False)
gdf_layers['union_cadastres_test'] = gdf_layers['union_cadastres_test'][gdf_layers['union_cadastres_test'].geometry.geom_type == "Polygon"]
print(f"Faster Overlay Identity took {time.perf_counter() - overlay_start:.2f} seconds")
```

Overlay Identity took 1228.24 seconds
Faster Overlay Identity took 119.04 seconds
A 10x speedup! and to confirm:

`gdf_layers['union_cadastres_test'].equals(gdf_layers['union_cadastres']) `
True

```
from geopandas.testing import assert_geodataframe_equal
assert_geodataframe_equal(gdf_layers['union_cadastres'], gdf_layers['union_cadastres_test'], 
                                      True, 'equiv','equiv',True,False,True,True,True,True)
```
No Error.

Shapely = 2.0.1
Geopandas = 0.14.4

**Comment by BMarques** [comment]
Sorry for not coming back in the meantime, I managed to find out another solution that solved my particular issue, but good to know about that.

**Comment by theroggy** [comment]
I implemented the speedup in https://github.com/geopandas/geopandas/pull/3504

**Comment by BMarques** [comment]
Would like to come back to this topic to give feedback, the enhancement in 1.1.1 compared to 1.0.0 was huge in my case, an identity went down from 2 hours to 20 minutes. Thank you.

**Comment by theroggy** [comment]
> Would like to come back to this topic to give feedback, the enhancement in 1.1.1 compared to 1.0.0 was huge in my case, an identity went down from 2 hours to 20 minutes. Thank you.

Nice, thanks for the feedback.

Depending on what you are aiming for... Just for information, as 20 minutes is still quite significant, you could have a look at [geofileops.identity](https://geofileops.readthedocs.io/en/stable/api/geofileops.identity.html) if you would like to try to speed it up some more...

**Comment by BMarques** [comment]
Sure, i'll give it a try, thanks
